iroha/src/block_sync.rs

```diff
@@ -0,0 +1,177 @@
+    pub fn start(&self) {
+        let gossip_period = self.gossip_period;
+        let kura = self.kura.clone();
+        let peer_id = self.peer_id.clone();
+        let sumeragi = self.sumeragi.clone();
+        task::spawn(async move {
+            loop {
+                task::sleep(gossip_period).await;
+                let message =
+                    Message::LatestBlock((kura.read().await.latest_block_hash(), peer_id.clone()));
+                futures::future::join_all(
+                    sumeragi
+                        .read()
+                        .await
+                        .network_topology
+                        .sorted_peers
+                        .iter()
+                        .map(|peer| message.clone().send_to(peer)),
+                )
+                .await;
+            }
+        });
+    }
+
+    /// Continues the synchronization if it was ongoing. Should be called after `WSV` update.
+    pub async fn continue_sync(&mut self) {
+        if let State::InProgress(blocks) = self.state.clone() {
+            if let Some((block, blocks)) = blocks.split_first() {
+                let mut network_topology = self.sumeragi.read().await.network_topology.clone();
+                network_topology.shift_peers_by_n(block.header.number_of_view_changes);
+                if self.kura.read().await.latest_block_hash() == block.header.previous_block_hash
+                    && network_topology
+                        .filter_signatures_by_roles(
+                            &[Role::ValidatingPeer, Role::Leader, Role::ProxyTail],
+                            &block.verified_signatures(),
+                        )
+                        .len()
+                        >= network_topology.min_votes_for_commit()
+                {
+                    self.state = State::InProgress(blocks.to_vec());
+                    self.sumeragi
+                        .write()
+                        .await
+                        .commit_block(block.clone())
+                        .await;
+                } else {
+                    self.state = State::Idle;
+                }
+            } else {
+                self.state = State::Idle;
+            }
+        }
+    }
+}
+
+/// The module for block synchronization related peer to peer messages.
+pub mod message {
+    use super::{BlockSynchronizer, State};
+    use crate::{block::ValidBlock, crypto::Hash, peer::PeerId, torii::uri};
+    use iroha_derive::*;
+    use iroha_network::prelude::*;
+    use parity_scale_codec::{Decode, Encode};
+
+    /// Message's variants that are used by peers to communicate in the process of consensus.
+    #[derive(Io, Decode, Encode, Debug, Clone)]
+    pub enum Message {
+        /// Gossip about latest block.
+        LatestBlock((Hash, PeerId)),
+        /// Request for blocks after the block with `Hash` for the peer with `PeerId`.
+        BlocksRequested((Hash, PeerId)),
+        /// Blocks.
+        Blocks(Vec<ValidBlock>),
```

Comment:
@eadventurous I think it will be even better to rename all variants into verb form:
* GetLatestBlock
* GetBlocksAfter(Hash)
* GetAllBlocks

Reply:
hm - my variants are wrong, maybe something like:
* NewLatestBlock
* ShareBlocksAfter
* CheckRequestedBlocks

Reply:
Will something like this be ok?
```rust
pub enum Message {
        /// Gossip about latest block.
        LatestBlock((Hash, PeerId)),
        /// Request for blocks after the block with `Hash` for the peer with `PeerId`.
        GetBlocksAfter((Hash, PeerId)),
        /// The response to `GetBlocksAfter`. Contains the requested blocks.
        ShareBlocks(Vec<ValidBlock>),
    }
```

Reply:
Just pushed the commit with this.